### ISSUE TITLE
Problem using correctionlib on CMSSW_12_X

### ISSUE THREAD

**Issue by GuillelmoGomezCeballos** [cms-nanoAOD/correctionlib#121]
Hi all,

I am trying to use correctionlib on CMSSW_12_2_X, and it's failing to me.

-- This works without any problem:

import ROOT
from correctionlib import _core
evaluator = _core.CorrectionSet.from_file("electron.json")
valsf = evaluator["UL-Electron-ID-SF"].evaluate("2018","sf","Medium",1.1,34.0)

-- This fails the last command:

import ROOT
import correctionlib
correctionlib.register_pyroot_binding()
ROOT.gInterpreter.Declare('auto csetEl = correction::CorrectionSet::from_file("electron.json");')
ROOT.gInterpreter.Declare('auto csetEl_2018preID = csetEl-at("UL-Electron-ID-SF");')
ROOT.gInterpreter.ProcessLine('csetEl_2018preID-evaluate({"2018","sf","Medium",1.1,34.0})')

In module 'ROOTNTuple':
/data/cmsbld/jenkins/workspace/auto-builds/CMSSW_12_1_0_pre5-slc7_amd64_gcc900/build/CMSSW_12_1_0_pre5-build/slc7_amd64_gcc900/external/gcc/9.3.0/bin/../lib/gcc/x86_64-unknown-linux-gnu/9.3.0/../../../../include/c++/9.3.0/variant:156:13: error: cannot cast private base class 'std::__detail::__variant::_Variant_storage<false, int, double, std::__cxx11::basic_string<char' to 'std::variant<int, double,
       std::__cxx11::basic_string<char'
             return static_cast<variant<_Types...&(__rhs);
...

I am trying to use these corrections within Root Data Frame, and for that I must be able to use the second way. Can this be fixed? I believe it's a rather important issue. Notice the json file type is not important, all of them suffer from the same problem.

Thanks, Guillelmo

**Comment by nsmith-** [comment]
Just to check, if you print out `repr(_core)` is it
```
>>> _core
<module 'correctionlib._core' from '/cvmfs/cms.cern.ch/slc7_amd64_gcc900/external/py3-correctionlib/2.0.0-0c4f44c8dd5561d8c0660135feeb81f4/lib/python3.9/site-packages/correctionlib/_core.cpython-39-x86_64-linux-gnu.so'>
```
i.e. coming from the cmssw installation?

Anyway I was able to reproduce the error, looking into it now

**Comment by GuillelmoGomezCeballos** [comment]
Yes, I believe so

Thanks for looking into it!

**Comment by nsmith-** [comment]
It seems it is a bug in ROOT older than v6.24 in conjunction with gcc9: https://sft.its.cern.ch/jira/browse/ROOT-10220
As simple as
```python
import ROOT
print(ROOT.__version__)
ROOT.gInterpreter.Declare('auto x = std::vector<std::variant<double,int,std::string>>({"hi", 3.1, 0});')
```
works in v6.24 but not v6.22, the latter is what is in CMSSW 12_2_X and 12_3_X it seems. I see two workarounds.

## Option 1:
Use ROOT v6.24 or newer, e.g. from LCG releases. For example, `/cvmfs/sft.cern.ch/lcg/views/LCG_101/x86_64-centos7-gcc11-opt/setup.sh` works, though you have to `pip3 install --user correctionlib --no-binary=correctionlib` since it is not part of LCG software package.
@pieterdavid is that how you usually work? Since your [examples](https://gist.github.com/pieterdavid/a560e65658386d70a1720cb5afe4d3e9) seem to rely on having this functionality. (the other possibility, if I understand the ROOT bug report correctly, is to be in an environment with gcc8 or lower)

## Option 2:
Wrap the needed corrections with your own class and pre-compile it so that the ROOT interpreter does not have to work with `std::variant`. For example, if I make `mysf.h`:
```cpp
#include "correction.h"


class MyCorrections {
  public:
    MyCorrections();
    double eval_electronSF(double pt, double eta);

  private:
    correction::Correction::Ref electronSF_;
};
```
and `mysf.cpp`:
```cpp
#include "mysf.h"

MyCorrections::MyCorrections() {
  auto csetEl = correction::CorrectionSet::from_file("electron.json");
  electronSF_ = csetEl->at("UL-Electron-ID-SF");
}

double MyCorrections::eval_electronSF(double pt, double eta) {
  return electronSF_->evaluate({"2018", "sf", "Medium", eta, pt});
};
```
then I can compile with `g++ $(correction config --cflags --ldflags) mysf.cpp -shared -fPIC -o mysf.so` and then the python script 
```python
import ROOT
import correctionlib

correctionlib.register_pyroot_binding()
ROOT.gInterpreter.Declare('#include "mysf.h"')
ROOT.gInterpreter.Load("mysf.so")
ROOT.gInterpreter.ProcessLine('auto corr = MyCorrections();')
ROOT.gInterpreter.ProcessLine('corr.eval_electronSF(34.0, 1.1)')
```
works inside CMSSW_12_2_0

**Comment by GuillelmoGomezCeballos** [comment]
Thanks for finding the problem!

**Comment by pieterdavid** [comment]
Yes, I've mostly used the LCG releases (and usually the most recent, so I'm not sure I've ever tried correctionlib with a pre-6.24 one), always installing from source in a virtualenv
